File: tournaments/combs/combs.js
function combs(comb1, comb2) {
    const getMask = function(comb) {
        let mask = 0;
        for (const item of comb) {
            mask <<= 1;
            mask += '*' === item;
        }

        return mask;
    };

    const m1 = getMask(comb1);
    const m2 = getMask(comb2);
    const len1 = comb1.length;
    const len2 = comb2.length;
    let result = len1 + len2;

    for (let i = -len1; len2 >= i; ++i) {
        let a = 0;
        let b = 0;

        if (0 > i) {
            a = m2 << (-i) & m1;
            b = Math.max(-i + len2, len1);
        } else {
            a = m1 << i & m2;
            b = Math.max(i + len1, len2);
        }
        if (!a && result > b) {
            result = b;
        }
    }

    return result;
}

File: tournaments/combs/combs.cpp
int combs(std::string comb1, std::string comb2) {

  struct Helper {
    int getMask(std::string const & comb) {
      int mask = 0;
      for (int i = 0; i < comb.size(); i++) {
        char c = comb[i];
        mask = (mask << 1) + (c == '*');
      }
      return mask;
    }
  };
  Helper h;

  int m1 = h.getMask(comb1);
  int m2 = h.getMask(comb2);
  int len1 = comb1.size();
  int len2 = comb2.size();
  int answer = len1 + len2;
  for (int i = -len1; i <= len2; i++) {
    int tmp, length;
    if (i < 0) {
      tmp = m2 << (-i) & m1;
      length = std::max(-i + len2, len1);
    } else {
      tmp = m1 << i & m2;
      length = std::max(i + len1, len2);
    }
    if (tmp == 0 && answer > length) {
      answer = length;
    }
  }

  return answer;
}
